File: leetcode/0907. Sum of Subarray Minimums/0907.py
class Solution:
  def sumSubarrayMins(self, arr: List[int]) -> int:
    kMod = int(1e9) + 7
    n = len(arr)
    ans = 0
    prev = [-1] * n
    next = [n] * n
    stack = []

    for i, a in enumerate(arr):
      while stack and arr[stack[-1]] > a:
        index = stack.pop()
        next[index] = i
      if stack:
        prev[i] = stack[-1]
      stack.append(i)

    for i, a in enumerate(arr):
      ans += a * (i - prev[i]) * (next[i] - i)
      ans %= kMod

    return ans

File: leetcode/0907. Sum of Subarray Minimums/0907.java
class Solution {
  public int sumSubarrayMins(int[] arr) {
    final int kMod = (int) 1e9 + 7;
    final int n = arr.length;
    long ans = 0;
    int[] prev = new int[n];
    int[] next = new int[n];
    Deque<Integer> stack = new ArrayDeque<>();

    Arrays.fill(prev, -1);
    Arrays.fill(next, n);

    for (int i = 0; i < arr.length; ++i) {
      while (!stack.isEmpty() && arr[stack.peek()] > arr[i]) {
        final int index = stack.pop();
        next[index] = i;
      }
      if (!stack.isEmpty())
        prev[i] = stack.peek();
      stack.push(i);
    }

    for (int i = 0; i < arr.length; ++i) {
      ans += (long) arr[i] * (i - prev[i]) * (next[i] - i);
      ans %= kMod;
    }

    return (int) ans;
  }
}
